File: 프로그래머스/unrated/154540. 무인도 여행/무인도 여행.py
import sys
sys.setrecursionlimit(10**5)

def dfs(maps, visited, x, y):
    R = len(maps)
    C = len(maps[0])
    
    visited[x][y] = True
    cnt = int(maps[x][y])
    
    for d in ((0, -1), (0, 1), (1, 0), (-1, 0)):
        nx = x + d[0]
        ny = y + d[1]
        
        if not (0<=nx<R and 0<=ny<C):
            continue
        
        if visited[nx][ny]:
            continue
        
        if maps[nx][ny] == "X":
            continue
            
        cnt += dfs(maps, visited, nx, ny)
    
    return cnt

def solution(maps):
    R = len(maps)
    C = len(maps[0])
    visited=[[False]*C for _ in range(R)]
    
    answer = []
    
    for i in range(R):
        for j in range(C):
            if maps[i][j] == "X":
                continue
            if not visited[i][j]:
                answer.append(dfs(maps, visited, i, j))
    answer.sort()
    
    if not answer:
        return [-1]
    else:
        return answer

File: 프로그래머스/2/154540. 무인도 여행/무인도 여행.java
import java.util.ArrayList;
import java.util.List;

class Solution {

    static int[] dx = {1, 0, -1, 0};
    static int[] dy = {0, 1, 0, -1};

    public int[] solution(String[] maps) {
        int N = maps.length;
        int M = maps[0].length();

        boolean[][] visited = new boolean[N][M];

        List<Integer> answer = new ArrayList<>();

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (maps[i].charAt(j) == 'X') {
                    continue;
                }

                if (visited[i][j]) {
                    continue;
                }

                answer.add(dfs(maps, i, j, visited));
            }
        }

        if (answer.isEmpty()) {
            return new int[]{-1};
        }
        return answer.stream().sorted().mapToInt(v -> v).toArray();
    }

    public int dfs(String[] maps, int x, int y, boolean[][] visited) {
        visited[x][y] = true;
        int cnt = Integer.parseInt(maps[x].charAt(y) + "");

        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];

            if (nx < 0 || nx >= maps.length || ny < 0 || ny >= maps[0].length()) {
                continue;
            }

            if (maps[nx].charAt(ny) == 'X') {
                continue;
            }

            if (visited[nx][ny]) {
                continue;
            }

            cnt += dfs(maps, nx, ny, visited);
        }

        return cnt;
    }
}